Samples/NetworkMediaStreamer/RTPMediaStreamer/src/RTPMediaSink.cpp

```diff
@@ -0,0 +1,211 @@
+            return S_OK;
+        }
+    }
+
+    STDMETHODIMP SetPresentationClock(
+        /* [in] */ __RPC__in_opt IMFPresentationClock* pPresentationClock) noexcept
+    {
+        RETURNIFSHUTDOWN;
+        if (m_spClock)
+        {
+            m_spClock->RemoveClockStateSink(this);
+            m_spClock = nullptr;
+        }
+        m_spClock.copy_from(pPresentationClock);
+        return m_spClock->AddClockStateSink(this);
+
+    }
+
+    STDMETHODIMP GetPresentationClock(
+        /* [out] */ __RPC__deref_out_opt IMFPresentationClock** ppPresentationClock) noexcept
+    {
+        RETURNIFSHUTDOWN;
+        if (!ppPresentationClock) return E_POINTER;
+        m_spClock.copy_to(ppPresentationClock);
+        return S_OK;
+    }
+
+    STDMETHODIMP Shutdown(void) noexcept
+    {
+        RETURNIFSHUTDOWN;
+        HRESULT hr = S_OK;
+        m_bIsShutdown = true;
+        for (auto strm : m_spStreamSinks)
+        {
+            auto hr1 = strm->Shutdown();
+            hr = FAILED(hr1)? hr1: hr;
+        }
+        return hr;
+    }
+
+    //IMediaExtension
+    STDMETHODIMP SetProperties(
+        winrt::Windows::Foundation::Collections::IPropertySet const& configuration) noexcept
+    {
+        return S_OK;
+    }
+
+    // IMFClockStateSink methods
+    STDMETHODIMP OnClockStart(MFTIME hnsSystemTime, LONGLONG llClockStartOffset) noexcept
+    {
+        HRESULT hr = S_OK;
+        for (auto strm : m_spStreamSinks)
+        {
+            auto hr1 = strm->Start(hnsSystemTime, llClockStartOffset);
+            hr = FAILED(hr1) ? hr1 : hr;
+        }
+        return hr;
+    }
+    STDMETHODIMP OnClockStop(MFTIME hnsSystemTime) noexcept
+    {
+        HRESULT hr = S_OK;
+        for (auto strm : m_spStreamSinks)
+        {
+            auto hr1 = strm->Stop(hnsSystemTime);
+            hr = FAILED(hr1) ? hr1 : hr;
+        }
+        return hr;
+    }
+    STDMETHODIMP OnClockPause(MFTIME hnsSystemTime) noexcept
+    {
+        HRESULT hr = S_OK;
+        for (auto strm : m_spStreamSinks)
+        {
+            auto hr1 = strm->Pause(hnsSystemTime);
+            hr = FAILED(hr1) ? hr1 : hr;
+        }
+        return hr;
+    }
+    STDMETHODIMP OnClockRestart(MFTIME hnsSystemTime) noexcept
+    {
+        HRESULT hr = S_OK;
+        for (auto strm : m_spStreamSinks)
+        {
+            auto hr1 = strm->Start(hnsSystemTime, 0);
+            hr = FAILED(hr1) ? hr1 : hr;
+        }
+        return hr;
+    }
+    STDMETHODIMP OnClockSetRate(MFTIME hnsSystemTime, float flRate) noexcept
+    {
+        return S_OK;
+    }
+};
+
+RTPMEDIASTREAMER_API STDMETHODIMP CreateRTPMediaSink(IMFMediaType** apMediaTypes, DWORD dwMediaTypeCount, IMFMediaSink** ppMediaSink)
+{
+    HRESULT_EXCEPTION_BOUNDARY_START;
+    winrt::check_pointer(ppMediaSink);
+    std::vector<winrt::com_ptr<IMFMediaType>> mediaTypes(dwMediaTypeCount);
+    for (DWORD i = 0; i < dwMediaTypeCount; i++)
+    {
+        mediaTypes[i].copy_from(apMediaTypes[i]);

```

Comment:
>copy_from [](start = 22, length = 9)

should this error on null?

Reply:
eventually it will error out during the construction of the NetworkVideoStreamSinkBase. But i think its a good idea to error out. over here..

---
In reply to: [491167716](https://github.com/microsoft/Windows-Camera/pull/13#discussion_r491167716) [](ancestors = 491167716)